### DISCUSSION TITLE
Проблема релинка в истории прослушиваний

### DISCUSSION THREAD

**Discussion by Chimildic** [Chimildic/goofy#99] [Принцип работы]
Релинк - это подмена видимого трека в момент воспроизведения на аналог с другим _id_, альбомом, иногда названием. Зачастую по причине недоступности оригинала в стране пользователя. Spotify совершает релинк настолько виртуозно, что не оставляет следов. Не смотря на документацию, описывающую его признаки.

Допустим в плейлисте есть трек `spotify:track:3fH4KjXFYMmljxrcGrbPj9`. При воспроизведении происходит релинк на `spotify:track:2u96akbqgCRtkjfLMYrgZo`. Spotify по запросу истории прослушиваний вернет второй трек. Поскольку слушали именно его. И никак не сообщит о подмене.

Классическое удаления истории из массива треков происходит через `Filter.removeTracks`. Совпадение трека происходит по _id_ и названию. Релинк даст новое _id_. Остается надеяться на совпадение по названию. Об этом узком месте пойдет речь. 

Когда название у релинк-трека и оригинала совпадает 1 к 1, трек удалится. Но если в одной из версий есть приписки типа `radio edition` или `mix` - трек останется. Становится неизвестно, произошел релинк или по-настоящему послушали другую версию трека? Именно это произошло с треками из примера.

> По своему опыту, не замечал подобной проблемы. Очень внимательным оказался пользователь [ilyal](https://4pda.to/forum/index.php?showuser=675488). В общем потоке, это крайне малое количество треков.

Есть решение, при котором меняется сравнение 1 к 1 на поиск подстроки в строке в обе стороны. При этом теряется смысл приписок. Неважно какую версию трека послушали, любая удалит любую.  Одностороннее сравнение будет полумерами.

| Версия | Плейлист | История | Удаляется  |
|-|-|-|-|
| Сейчас | Don't You (Forget About Me) | Don't You (Forget About Me) | Да |
|| Don't You (Forget About Me)  | Don't You (Forget About Me) - 2001 mix | Нет |
|| Don't You (Forget About Me) - 2001 mix | Don't You (Forget About Me) | Нет |
| Одностороннее| Don't You (Forget About Me) | Don't You (Forget About Me) | Да |
|| Don't You (Forget About Me)  | Don't You (Forget About Me) - 2001 mix | Нет |
|| Don't You (Forget About Me) - 2001 mix | Don't You (Forget About Me) | Да |
| Двухстороннее | Don't You (Forget About Me) | Don't You (Forget About Me) | Да |
|| Don't You (Forget About Me)  | Don't You (Forget About Me) - 2001 mix | Да |
|| Don't You (Forget About Me) - 2001 mix | Don't You (Forget About Me) | Да |

Как видно из таблицы, _сейчас_ работает правильно. Оригинал и микс - разные треки. Только присутствует кот Шредингера в лице релинка. Была подмена или нет?

Так как это не массовое явление, предлагаю ограничиться альтернативами:
- если до сих пор подобное не замечено или не мешает, ничего не делать.
- заменить трек в плейлисте на тот, который приходит с релинка. 
- если допустимо игнорировать неоригинальные версии, использовать фильтр `match*` вместе с удалением истории.
- и наконец, занесите релинк-трек в историю руками [appendTracks](https://chimildic.github.io/goofy/#/func?id=appendtracks). Поскольку история постепенно заменяется новыми треками, имеет смысл создать отдельный кэш файл, чтобы в будущем быстрее занести такие треки в историю. Удобным будет организовать [форму для телефона](https://github.com/Chimildic/goofy/discussions/9).

**Comment by ilyanikolaich** [comment]
Как первый замеченный наблюдатель этой "фичи" внесу своё мнение. Этот эффект проявляется крайне редко, лично я сталкивался с ним (по крайней мере, замечено и осознанно) - раза три наверное. Пару раз разводил руками удивляясь, но вот на третий удалось поймать и локализовать суть вопроса.
Поскольку это довольно редкое явление, то моё мнение - ничего в алгоритмах менять не надо. Данная ситуация интересна тем, что позволила "отловить" и объяснить подобное поведение алгоритмов. То есть, за счет своей крайне редкой частоты появления предлагаю обозначить это "не багом, а фичей", зафиксировать (что уже и сделано) в обсуждениях, дабы в будущем было доступно быстрое и логичное объяснение этого эффекта.

Для себя решил вопрос немного костыльно, но чем дальше, тем мне такой вариант нравится больше. А именно - у меня уже давно присутствует плейлист, в который временно попадают треки, которые я по каким-то личным причинам слушать не хочу. Вот прямо сейчас - не хочу и всё. Это не значит, что мне не нравится исполнитель, или даже сам трек. Именно поэтому для меня этот мой TempBanList - временный. Наверное раз в пару-тройку недель под настроение я его пересматриваю и подчищаю. Трек с замеченным неконтролируемым релинком - я поместил туда. И к полной истории прослушивания я добавляю треки этого плейлиста. Такое поведение кода меня устраивает - оно мне понятно и я могу его регулировать.

В очередной раз огромная благодарность Chimildic за разбор полётов "посимвольно"

**Comment by ilyanikolaich** [comment]
@Chimildic 
Вот попался красивый образец релинкового беспорядка.
Скриншот из десктоп клиента Спотифай:

![image](https://user-images.githubusercontent.com/60379082/124701376-ac6d8580-df18-11eb-9a04-d672f78af5e0.png)
Первые шесть треков (выделены) - прослушаны полностью, но попали в историю прослушивания под релинковыми "копиями". То есть реально прослушаны были как раз те самые копии. А эти треки после очистки истории - остались на месте. То есть это все полностью совпадает с вышеописанным эффектом.

Параллельно обрисовался еще один :
Если **этот же плейлист** открыть в веб версии Спотифай в браузере, то - о чудо! - мы увидим те треки, которые реально были прослушаны, и которые зафиксировались в истории прослушивания

Скриншот из веб-версии Спотифай
![image](https://user-images.githubusercontent.com/60379082/124701868-82689300-df19-11eb-8a57-283998dd23e9.png)

Я не удержался, и произвел повторное прослушивание этих шести треков через веб интерфейс, и попробовал очистить историю прослушивания плейлиста. Нет, еще одного чуда не произошло - эти треки так и остались на месте. То есть, в данном случае по этим шести трекам - в десктоп-клиенте и в приложении на андроид - отображаются треки, которые в реальности линкуются на другие, но по каким-то причинам в веб версии плеера - отображаются те, которые будут реально играть. Чудеса вокруг нас.

Скриншот приложения андроид:
![image](https://user-images.githubusercontent.com/60379082/124702428-8c3ec600-df1a-11eb-9a16-132c681c467c.png)

Дополнительно интересный факт в этой всей ситуации с треком Алисы - Мы вместе. По какой причине этот трек не отфильтровался, я не могу понять. Вот его данные
```
есть в плейлисте Time Capsule:
Алиса - Мы вместе
spotify:track:78SgVXMfDoNS8TGEU2JW5L
spotify:album:6zUr1rWc9Fs7ecpVVG2JjD

Попало в историю, и не удаляется из изначального листа:
Алиса - Мы вместе
spotify:track:1quj5b89I1d4cUkOBXnE5m
spotify:album:5EIB9CV6uGDDKcIWQww65B
```

**Reply by Chimildic** [reply]
> факт в этой всей ситуации с треком Алисы

У Spotify есть фишка. По API давать треки с разным написанием имен. Если не просить _явно_ кириллицу: одни будут приходить всегда кириллицей, другие придут на английской транскрипции. Причем вразнобой: все транскрипцией или только исполнитель, или только трек.

Чтобы всегда иметь кириллицу, в одной из последних итераций версии 1.4.8 к запросу треков из плейлиста явно добавлено `RU`. Помогает с фильтрами по `match`.

Прослушал трек Алисы, в историю пришло микс название. Поэтому при удалении встречаются `Alisa - Мы вместе` из истории и `Алиса - Мы вместе` из плейлиста. При условии, что у вас последнее обновление с `RU`, о котором выше сказано. В предыдущих версиях сравнивались бы микс-названия. Но при этом был сломанный `match` по кириллице.

Подставьте в следующий код _id_ от плейлиста капсулы. Увидите какое написание у этих треков в данный момент.
```js
function debug() {
  let playlistTrack = Source.getPlaylistTracks('', 'id капсулы').find(t => t.id == '78SgVXMfDoNS8TGEU2JW5L');
  let historyTrack = RecentTracks.get().find(t => t.id == '1quj5b89I1d4cUkOBXnE5m');

  console.log('В плейлисте: ', `${playlistTrack.artists[0].name} - ${playlistTrack.name}`);
  console.log('В истории:', `${historyTrack.artists[0].name} - ${historyTrack.name}`);
}
```

> При этом [онлайн консоль Spotify](https://developer.spotify.com/console/get-recently-played) дает полную кириллицу при запросе истории прослушиваний. Может быть он на IP ссылается, когда нет явных указаний. 

В общем, добавил явно `RU` и к запросу истории, пришла полная кириллица. Так что багофича может быть исправлена. Чуть позже обновлю библиотеку.

**Reply by ilyanikolaich** [reply]
Да, всё верно:
```
14:01:03	Информация	В плейлисте:  Алиса - Мы вместе
14:01:03	Информация	В истории: Alisa - Мы вместе
```
